amendment to proposal amendment (#189)
cylc/cylc-admin: docs/proposal-optional-output-extension.md

   executed and that might not happen if expiry or submission are optional (or if the pre-requisites
   for `a` are not met).

> 3. A task is complete if the completion condition is satisfied.

Author: ```suggestion
3. Outputs are complete if the completion condition is satisfied.
```

Author: @hjoliver is this correction sufficient for you?

Reviewer: Not by itself - it has to be carried throughout the document wherever "complete tasks" and "incomplete tasks" are referred to.

I take it you are agreeing to [2. above](https://github.com/cylc/cylc-admin/pull/189#issuecomment-2010574702) then?

Reviewer: Happy to have a crack at that myself if you're busy on implementation?

Author: > I take it you are agreeing to https://github.com/cylc/cylc-admin/pull/189#issuecomment-2010574702 then?

No, I'm sticking with:

* Outputs are complete if the completion expression is satisfied.
* Tasks are complete if they have a final task status AND their outputs are complete.

Reminder, `TaskOutputs.is_complete` is an internal interface not exposed to users, this definition is not important.

Providing there aren't any technical qualms with this, I'm going to plug on with implementation as there's still quite a lot to do.

Reviewer: But are you also dropping your proposed definition of *incomplete tasks*? If not, we're at an impasse. 

My bottom line is we can't redefine the terms complete and incomplete so that `"not complete" != "incomplete"`

Yes this isn't technical in the sense that we agree entirely on the logic, but it sure would be nice if the implementation didn't use inconsistent terms too.

Reviewer: > eh? Tasks and TaskOutputs are two different things. I see no issue with what is written.

Your single-line change does not address my concerns without additional information.  As far as I can tell you are now saying that you won't call tasks complete if their outputs are complete, but you will still call them incomplete if their outputs are incomplete AND they are finished. That's a bit of a mess compared to either of my two ways to fix the inconsistent terminology.

You could stop spending time on this by simply agreeing - this is a review of a proposal and I've raised a valid objection to the terminology being used.

Author: As far as I can tell you are now saying that you won't call tasks complete if their outputs are complete, but you will still call them incomplete if their outputs are incomplete AND they are finished

Correct.

> You could stop spending time on this by simply agreeing

Harsh. Your turn to propose something, my turn to be nit picky.

Reviewer: Sorry, didn't mean to be harsh - but I have proposed a solution (two of them in fact), in great detail, above. This is *only* a small question of consistent use of two common terms, it doesn't require a whole new proposal. Are you sure you've read and understood my "proposals"? (This sort of thing would be a lot easier to resolve if we didn't live in opposite timezones!)

Reviewer: How about I merge this and immediately put up a tweak PR that redoes the use of "in/complete tasks" in the way that I propose?  (However, don't take that as full approval of this one yet :grin:)

Author: If you could pick an optional and raise it so I can review it in context.

Reviewer: Ok, I'll clone your branch (rather than merge it) and put up an alternative.

   The completion condition can be defined by the user, otherwise the
   expression is automatically generated according to the rules:

>    1. All required outputs (referenced in the graph) must be satisfied.
>    2. If success is optional, then the task is complete if it fails.
>    3. If submission is optional, then the task is complete if it submit-fails.
>    4. If expiry is optional, then the task is complete if it expires.

Reviewer: ```suggestion
   1. If all required outputs (referenced in the graph) are satisfied, then the task is complete if it is finished.
   2. If success is optional, then the task is complete if it fails.
   3. If submission is optional, then the task is complete if it submit-fails.
   4. If expiry is optional, then the task is complete if it expires.
```

Reviewer: Also, we need to decide whether outputs get "completed" or "satisfied". I have always said the former (only prerequisites get satisfied). However I suppose we could say that outputs get satisfied, in order to avoid any confusion with task completion.

Author: > These points need ORs

Can add `or`s, it doesn't change the meaning.

> and for 1., the task must be finished, right?

No, we determine whether a task is "finished" from the task state not the task outputs. E.G. a retrying task may have the "failed" task output but be in the "waiting" state. Attempting to build the "finished" check into the completion expression results in bugs.

The completion expression tells you whether a task is complete, a task is *in*complete if this expression evaluates `False` AND the task is in a final *state*.

Author: ```suggestion
   1. All required outputs (referenced in the graph) must be satisfied.
   2. Or, if success is optional, then the task is complete if it fails.
   3. Or, if submission is optional, then the task is complete if it submit-fails.
   4. Or, if expiry is optional, then the task is complete if it expires.
```

Reviewer: > No, we determine whether a task is "finished" from the task state not the task outputs...

Yes but no but. By "finished" I did mean a final state, not outputs. 

I missed this critical change a bit higher up, sorry, which partly explains your response:
```diff
- 3. A task is incomplete if the completion condition is not satisfied and:
+ 3. A task is complete if the completion condition is satisfied.
```

This is change in terminology is not ideal, at least not without other attendant changes. 

Currently:
- *a task is complete **if** its outputs are complete **and** it is finished
- *a task is incomplete **if** its outputs are not complete **and** it is finished

[N.B., by "its outputs are complete" I mean either "its required outputs are complete" (currently) or "its completion condition is satisfied" (soon).]

So currently the word "complete" is a bit overloaded, but at least complete and incomplete tasks  are logical opposites. And that's important because it describes when the scheduler can forget (as complete) or retain (as incomplete) a finished task.

If we change to *a task is complete if its output completion condition is satisfied* (even if it's not finished), then logically a task should be incomplete if its output completion condition is NOT satisfied (even if not finished). But we still have *a task is incomplete, if the completion condition is not satisfied but the task has a final task status*.

I think we need to change the definition of incomplete too:
- a task is complete if its output completion condition is satisfied
- a task is incomplete if its output completion condition is not satisfied
- a finished task will be removed if complete, and retained if incomplete

Reviewer: Following up on that, let's go with your use of "satisfied" instead of "completed" for individual outputs as well. We've always said that prerequisites/dependencies get satisfied and outputs get completed, which makes sense, but in the new world perhaps its OK to say that "task outputs get satisfied by task messages (or by `cylc set`). That will remove all ambiguity from the concept of "completion" in Cylc:

- individual outputs get satisfied (not completed) by task messages or `cylc set`
- a task is complete if its output completion condition is satisfied, otherwise it is incomplete
- finished tasks get removed from the active pool if complete, or retained if incomplete

OK?

Reviewer: (I think that's my only problem with the amended amended proposal).

Author: > a task is incomplete if its outputs are not complete and it is finished

Read down the proposal a couple lines, this is already in there.

Author: > let's go with your use of "satisfied" instead of "completed" for individual outputs

I'm not fussed by grammar gerrymandering. I would lean towards:

* An expression can be "satisfied".
* A task output can be "completed".

If you want something different let me know.

Author: > I think that's my only problem with the amended amended proposal, otherwise great

I think the issue is grammar? Assuming there's no fundamental issue with this proposal, I'm going to push ahead with the branch in the name of haste.

Reviewer: > >    a task is incomplete if its outputs are not complete and it is finished

> Read down the proposal a couple lines, this is already in there.

I know. My point is: that is now logically inconsistent with your new definition of "completed task".

> Sorry, I'm a bit confused about what you're suggesting. This proposal states:
> >    A task is complete if the completion condition is satisfied.
>>  And:
> >   a task is incomplete, if the completion condition is not satisfied
    but the task has a final task status (note status not output). This is the
    same as presently implemented (8.2.x)
>
> Ok?

I agree (well, I observe) that the proposal states both of those things, but the problem is that means "complete" and "incomplete" tasks are no longer logically opposite - only one of them cares about final task status - and I think that is a problem for clarity and user understanding.

> >  let's go with your use of "satisfied" instead of "completed" for individual outputs
> I'm not fussed by grammar gerrymandering. I would lean towards:
>    - An expression can be "satisfied".
>    - A task output can be "completed".
> If you want something different let me know.

This isn't grammar, it's about the definition of the words that we're using to describe these concepts.  In this case, I also lean towards those existing uses, but I thought **you** wanted something different! I was literally quoting **your wording** in the proposal ("satisfying" rather than "completing" an output), assuming that was deliberate, and speculating that your rationale was to disambiguate our current overloaded use of the word "complete".

> I think the issue is grammar? Assuming there's no fundamental issue with this proposal, I'm going to push ahead with the branch in the name of haste. Let me know if I've misinterpreted this.

It's not grammar, as I said, it's about defining terms clearly and consistently. But yes, that's still not a fundamental issue, so please do go ahead with implementation (sorry I was under the impression you already were working on it, and didn't realize you might think this was a "fundamental" objection).

   }
   ```

   Note: the completion condition only covers outputs generated when a task is executed - it cannot
   include `:expired`, `:submit` or `:submit-fail`. It also cannot include `started` (this has to

>    happen if a task executes) or `finished` (this isn't an output).

Author: This restriction is no longer necessary with the above change.

We could decide to preserve this restriction, however, it will be clearer for users if we broaden the scope of the completion condition to include pre-conditions as otherwise we will have to automatically add the pre-condition logic internally.
